internal/compiler/typeregister.rs

```diff
@@ -65,14 +65,110 @@ macro_rules! declare_enums {
                     );
                 })*
             }
+            fn get_enum(&self, name: &str) -> Type {
+                match name {
+                    $( stringify!($Name) => Type::Enumeration(self.$Name.clone()), )*
+                    _ => unreachable!()
+                }
+            }
         }
     };
 }
 
 i_slint_common::for_each_enums!(declare_enums);
 
+macro_rules! declare_export_structs {
+    (
+        // match `KeyboardModifiers` first since it's dependent by following struct instances
+        $(#[$attr0:meta])*
+        struct $Name0:ident {
+            @name = $inner_name0:literal
+            export {
+                $( $(#[$pub_attr0:meta])* $pub_field0:ident : $pub_type0:ty, )*
+            }
+            private {
+                $( $(#[$pri_attr0:meta])* $pri_field0:ident : $pri_type0:ty, )*
+            }
+        }
+
+        $(
+            $(#[$attr:meta])*
+            struct $Name:ident {
+                @name = $inner_name:literal
+                export {
+                    $( $(#[$pub_attr:meta])* $pub_field:ident : $pub_type:ty, )*
+                }
+                private {
+                    $( $(#[$pri_attr:meta])* $pri_field:ident : $pri_type:ty, )*
+                }
+            }
+        )*
+    ) => {
+        pub struct ExportStructs {
+            pub $Name0: Type,
+            $(pub $Name: Type),*
+        }
+        impl ExportStructs {
+            fn new() -> Self {
+                let $Name0 = Type::Struct {
+                    fields: BTreeMap::from([
+                        $((stringify!($pub_field0).replace('_', "-"), match stringify!($pub_type0) {
+                            "bool" => Type::Bool,
+                            "i32" => Type::Int32,
+                            "f32" => Type::Float32,
+                            "SharedString" => Type::String,
+                            "Coord" => Type::LogicalLength,
```

Comment:
Maybe here, instead of runtime matching, it can be compile time with a macro.

something like
```
macro_rules! map_type {
  (bool) => { Type::Bool };
  (i32) => { Type::Int32 };
  ...
}
```

Reply:
Thanks for the hint!

After a bit of refactoring, finally I eliminated special match for `KeyboardModifiers`, make the macro `declare_export_structs` general declaration for builtin structs.

Please have a look one more time, appreciate!